src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs

```diff
@@ -103,33 +101,57 @@ private static bool IsDeadProducer(IStreamProducerExtension producer)
 
         public async Task<ISet<PubSubSubscriptionState>> RegisterProducer(StreamId streamId, IStreamProducerExtension streamProducer)
         {
-            if (!IsActiveProducer(streamProducer))
-                throw new ArgumentException(String.Format("Trying to register non active IStreamProducerExtension: {0}", streamProducer.ToString()), "streamProducer");
-            
-            RemoveDeadProducers();
-            
-            var publisherState = new PubSubPublisherState(streamId, streamProducer);
-            State.Producers.Add(publisherState);
             counterProducersAdded.Increment();
-            counterProducersTotal.Increment();
-            LogPubSubCounts("RegisterProducer {0}", streamProducer);
-            await WriteStateAsync();
+            if (!IsActiveProducer(streamProducer))
+                throw new ArgumentException($"Trying to register non active IStreamProducerExtension: {streamProducer}", "streamProducer");
+
+            try
+            {
+                RemoveDeadProducers();
+
+                var publisherState = new PubSubPublisherState(streamId, streamProducer);
+                State.Producers.Add(publisherState);
+                LogPubSubCounts("RegisterProducer {0}", streamProducer);
+                await WriteStateAsync();
+                counterProducersTotal.Increment();
+            }
+            catch (Exception exc)
+            {
+                logger.Error(ErrorCode.Stream_RegisterProducerFailed, $"Failed to register a stream producer.  Stream: {streamId}, Producer: {streamProducer}, Exception: {exc}");
+                // Inconsistent state, deactivate grain.
```

Comment:
> "Why do you think its Inconsistent state?"

By this point in the call we've modified the grain state, but received an error (any error) when writing to storage, so we assume inconsistent state because we've no guarantee that the grain state and the state in storage are the same.

> "I think the design was that is grain does not handle the explicit Inconsistent state exc, the runtime will deactivate"

Yes, but that is not the current behavior, and addressing that effects the way grain calls behave. This PR is narrowly focused on hardening this grain, not addressing the wider holes of the Orleans storage provider model.